src/Chapter3.hs

```diff
@@ -603,28 +598,33 @@ countPeople house = case people house of
     ThreePersons -> 3
     FourPersons -> 4
 
+makeHouse :: Int -> Maybe House
+makeHouse numPersons = case numPersons of
+    1 -> Just (House OnePerson)
+    2 -> Just (House TwoPersons)
+    3 -> Just (House ThreePersons)
+    4 -> Just (House FourPersons)
+    _ -> Nothing
+
 buildHouse  :: City -> Int -> City
-buildHouse city numPersons = city { houses = (new_house:houses city)}
-    where
-        new_house
-            | numPersons == 1 = House OnePerson
-            | numPersons == 2 = House TwoPersons
-            | numPersons == 3 = House ThreePersons
-            | numPersons == 4 = House FourPersons
-            | otherwise = error "Invalid number of persons"
-
--- Do not know exactly how to check that a city has a castle
-buildWalls :: City -> City
-buildWalls city
-    | population >= 10
---      I tried this approach, but didn't make it work out
---      && (typeOf city == CityWithCastleWallChurch
---      ||  typeOf city == CityWithCastleWallLibrary )
-      = city { walls=(Wall:walls city) }
-    | otherwise = error "Not enough population :("
+buildHouse city numPersons = case new_house of
+    Just house -> city { cityHouses = (house:cityHouses city)}
+    Nothing -> city
     where
-        population :: Int
-        population = sum (map countPeople (houses city))
+        new_house = makeHouse numPersons
+
+---- Do not know exactly how to check that a city has a castle
```

Comment:
You can use `case-of` here:

```haskell
... = case cityCastle city of
    Castle castle -> city { cityCastle = WallCastle castle}
    _ -> city
```

So you actually need to do something only when there is castle without walls, in all other cases, nothing could be done 🙂